CI: Spark download 404s from dlcdn.apache.org

## What

Every recent CI run fails at the \"Setup Apache Spark\" step with:

~~~
+ wget -q https://dlcdn.apache.org/spark/spark-4.0.0/spark-4.0.0-bin-hadoop3.tgz
##[error]Process completed with exit code 8.
~~~

wget exit code 8 is server-returned error — the file isn't there. \`dlcdn.apache.org\` only keeps current releases; older (and occasionally even current) versions 404 intermittently as the mirror rotates.

## Why

The download is the only step in the workflow that pulls from the internet at runtime. Every other dependency is cached or resolved through Go's module proxy. One transient 404 blocks every PR from getting a green CI, so the signal stops being useful.

## Proposal

Try \`dlcdn.apache.org\` first (faster, closer mirrors) and fall back to \`archive.apache.org\` (canonical, never rotates) on 404. Both host identical artefacts for every released Spark version.

~~~bash
wget -q https://dlcdn.apache.org/spark/spark-${VER}/spark-${VER}-bin-hadoop${H}.tgz || \
wget -q https://archive.apache.org/dist/spark/spark-${VER}/spark-${VER}-bin-hadoop${H}.tgz
~~~

The cache step already keyed on version + hadoop version means this penalty is paid at most once per cache eviction.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -84,7 +84,12 @@ jobs:
           echo "Apache Spark is not installed"
           # Access the directory.
           mkdir -p ~/deps/
-          wget -q https://dlcdn.apache.org/spark/spark-${{ env.SPARK_VERSION }}/spark-${{ env.SPARK_VERSION }}-bin-hadoop${{ env.HADOOP_VERSION }}.tgz
+          # dlcdn.apache.org only keeps current releases on its mirrors and
+          # occasionally 404s on older ones. archive.apache.org is the
+          # canonical mirror and never rotates — use it as a fallback.
+          ARCHIVE=spark-${{ env.SPARK_VERSION }}-bin-hadoop${{ env.HADOOP_VERSION }}.tgz
+          wget -q https://dlcdn.apache.org/spark/spark-${{ env.SPARK_VERSION }}/$ARCHIVE || \
+          wget -q https://archive.apache.org/dist/spark/spark-${{ env.SPARK_VERSION }}/$ARCHIVE
           tar -xzf spark-${{ env.SPARK_VERSION }}-bin-hadoop${{ env.HADOOP_VERSION }}.tgz -C ~/deps/
           # Delete the old file
           rm spark-${{ env.SPARK_VERSION }}-bin-hadoop${{ env.HADOOP_VERSION }}.tgz

--- a/spark/client/client_test.go
+++ b/spark/client/client_test.go
@@ -19,6 +19,9 @@ import (
 	"bytes"
 	"context"
 	"errors"
+	"testing"
+	"time"
+
 	"github.com/apache/arrow-go/v18/arrow"
 	"github.com/apache/arrow-go/v18/arrow/array"
 	"github.com/apache/arrow-go/v18/arrow/ipc"
@@ -31,8 +34,6 @@ import (
 	"github.com/google/uuid"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"testing"
-	"time"
 )
 
 func TestAnalyzePlanCallsAnalyzePlanOnClient(t *testing.T) {

--- a/spark/sql/types/rowiterator.go
+++ b/spark/sql/types/rowiterator.go
@@ -3,10 +3,11 @@ package types
 import (
 	"context"
 	"errors"
-	"github.com/apache/arrow-go/v18/arrow"
 	"io"
 	"sync"
 	"time"
+
+	"github.com/apache/arrow-go/v18/arrow"
 )
 
 // RowIterator provides streaming access to individual rows
@@ -116,7 +117,6 @@ func (iter *rowIteratorImpl) fetchNextBatch() error {
 				defer record.Release()
 				return ReadArrowRecordToRows(record)
 			}()
-
 			if err != nil {
 				return err
 			}
@@ -146,7 +146,6 @@ func (iter *rowIteratorImpl) fetchNextBatch() error {
 						defer record.Release()
 						return ReadArrowRecordToRows(record)
 					}()
-
 					if err != nil {
 						return err
 					}

--- a/spark/sql/types/rowiterator_test.go
+++ b/spark/sql/types/rowiterator_test.go
@@ -3,11 +3,12 @@ package types_test
 import (
 	"context"
 	"errors"
-	"github.com/apache/arrow-go/v18/arrow/array"
-	"github.com/apache/arrow-go/v18/arrow/memory"
 	"io"
 	"testing"
 	"time"
+
+	"github.com/apache/arrow-go/v18/arrow/array"
+	"github.com/apache/arrow-go/v18/arrow/memory"
 
 	"github.com/apache/arrow-go/v18/arrow"
 	"github.com/stretchr/testify/assert"

